drivers/usb/udc/udc_smartbond.c

```diff
@@ -0,0 +1,1749 @@
+	rx_block->dest_addr_adj = 0x0;
+
+	/* Should reflect USB RX FIFO */
+	rx_block->source_address = 0;
+	/* Should reflect RX buffer. Temporarily assign an SRAM location. */
+	rx_block->dest_address = MCU_SYSRAM_M_BASE;
+	/* Should reflect total bytes to be received */
+	rx_block->block_size = 0;
+
+	if (dma_config(dma_cfg->rx_dev, dma_cfg->rx_chan, rx) < 0) {
+		LOG_ERR("RX DMA configuration failed");
+		return -EINVAL;
+	}
+
+	if (dma_config(dma_cfg->tx_dev, dma_cfg->tx_chan, tx) < 0) {
+		LOG_ERR("TX DMA configuration failed");
+		return -EINVAL;
+	}
+
+	return 0;
+}
+
+static void usb_smartbond_dma_deconfig(struct usb_smartbond_data *data)
+{
+	const struct udc_smartbond_config *config = data->dev->config;
+	const struct usb_smartbond_dma_config *dma_cfg = &config->dma_cfg;
+
+	dma_stop(dma_cfg->tx_dev, dma_cfg->tx_chan);
+	dma_stop(dma_cfg->rx_dev, dma_cfg->rx_chan);
+
+	dma_release_channel(dma_cfg->tx_dev, dma_cfg->tx_chan);
+	dma_release_channel(dma_cfg->rx_dev, dma_cfg->rx_chan);
+}
+
+static struct smartbond_ep_state *usb_dc_get_ep_state(struct usb_smartbond_data *data, uint8_t ep)
+{
+	const struct udc_smartbond_config *config = data->dev->config;
+
+	uint8_t ep_idx = USB_EP_GET_IDX(ep);
+	uint8_t ep_dir = USB_EP_GET_DIR(ep) ? 1 : 0;
+
+	return (ep_idx < config->num_of_eps) ? &data->ep_state[ep_dir][ep_idx] : NULL;
+}
+
+static struct smartbond_ep_state *usb_dc_get_ep_out_state(struct usb_smartbond_data *data,
+							  uint8_t ep_idx)
+{
+	const struct udc_smartbond_config *config = data->dev->config;
+
+	return ep_idx < config->num_of_eps ? &data->ep_state[0][ep_idx] : NULL;
+}
+
+static struct smartbond_ep_state *usb_dc_get_ep_in_state(struct usb_smartbond_data *data,
+							 uint8_t ep_idx)
+{
+	const struct udc_smartbond_config *config = data->dev->config;
+
+	return ep_idx < config->num_of_eps ? &data->ep_state[1][ep_idx] : NULL;
+}
+
+static void set_nfsr(struct usb_smartbond_data *data, uint8_t val)
+{
+	data->nfsr = val;
+	/*
+	 * Write only lower 2 bits to register, higher bits are used
+	 * to count down till OPERATIONAL state can be entered when
+	 * remote wakeup activated.
+	 */
+	USB->USB_NFSR_REG = val & 3;
+}
+
+static void fill_tx_fifo(struct smartbond_ep_state *ep_state)
+{
+	int remaining;
+	const uint8_t *src;
+	struct smartbond_ep_reg_set *regs = ep_state->regs;
+	struct net_buf *buf = ep_state->buf;
+	const struct udc_ep_config *const ep_cfg = &ep_state->config;
+	const uint16_t mps = udc_mps_ep_size(ep_cfg);
+	const uint8_t ep_idx = USB_EP_GET_IDX(ep_cfg->addr);
+
+	src = buf->data;
+	remaining = buf->len;
+	if (remaining > mps - ep_state->last_packet_size) {
+		remaining = mps - ep_state->last_packet_size;
+	}
+
+	/*
+	 * Loop checks TCOUNT all the time since this value is saturated to 31
+	 * and can't be read just once before.
+	 */
+	while ((regs->txs & USB_USB_TXS1_REG_USB_TCOUNT_Msk) > 0 && remaining > 0) {
+		regs->txd = *src++;
+		ep_state->last_packet_size++;
+		remaining--;
+	}
+
+	if (ep_idx != 0) {
```

Comment:
Wondering why you do not need to handle it for control endpoints?
Also, for constructs like this, you can just reverse the logic and just return. The code below will look a bit more relaxed.

```c
	if (ep_idx == 0) {
		/* Not needed for control endpoints because they are fancy. */
		return;
	}
```

Reply:
Suggestion applied. Added explanation why this case does not apply. Additional check also added for cases where enabling FIFO level interrupt may not be needed.